### PR TITLE
8336999: Verification for resource area allocated data structures in C2

### PR DESCRIPTION
**Pull request by TobiHartmann**
Similar to [GrowableArrayNestingCheck](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/utilities/growableArray.cpp#L60), we should implement a check for C2's resource allocated data structures that verifies that reallocation happens under the same `ResourceMark` as the original allocation. Otherwise, use-after-free bugs like [JDK-8336095](https://bugs.openjdk.org/browse/JDK-8336095) will lead to memory corruption.

This change adds a [ReallocMark](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/allocation.cpp#L233) to all resource allocated data structures used by C2. I slightly modified it such that it checks the arena and skips verification if the data is not allocated in the resource arena. I also modified the grow methods such that we perform verification even if no reallocation is required. In addition, I changed a few `Unique_Node_List` allocations in vector.cpp from `comp_arena` to resource area allocations because they only have a short lifetime.

While testing, I hit the verification code from:
```
V  [libjvm.so+0x5c1ceb]  ReallocMark::check(Arena*)+0x7b  (allocation.cpp:244)
V  [libjvm.so+0x6df2da]  Block_Array::grow(unsigned int)+0x1a  (block.cpp:43)
V  [libjvm.so+0xb88679]  PhaseCFG::do_DFS(Tarjan*, unsigned int)+0x159  (block.hpp:72)
V  [libjvm.so+0xb88b6b]  PhaseCFG::build_dominator_tree()+0xab  (domgraph.cpp:74)
V  [libjvm.so+0xd75791]  PhaseCFG::do_global_code_motion()+0x11  (gcm.cpp:1635)
V  [libjvm.so+0x9f4fd4]  Compile::Code_Gen()+0x2a4  (compile.cpp:2949)
V  [libjvm.so+0x9f5f16]  Compile::Compile(ciEnv*, TypeFunc const* (*)(), unsigned char*, char const*, int, bool, bool, DirectiveSet*)+0xba6  (compile.cpp:991)
```

It's a false positive because the code in `PhaseCFG::build_dominator_tree` pre-grows `PhaseCFG::_blocks` to prevent reallocation before entering the scope of a nested ResourceMark. I think that's bad practice and should be avoided. I changed the code to allocate `_blocks` in a separate arena and removed the pre-growing.

This detects [JDK-8336095](https://bugs.openjdk.org/browse/JDK-8336095) right away, even with `java -Xcomp -version`.

We should revisit the footprint impact of arena allocations in C2 with [JDK-8337015](https://bugs.openjdk.org/browse/JDK-8337015).

Thanks,
Tobias

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336999](https://bugs.openjdk.org/browse/JDK-8336999): Verification for resource area allocated data structures in C2 (**Enhancement** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) 🔄 Re-review required (review applies to [b0f839b8](https://git.openjdk.org/jdk/pull/20311/files/b0f839b898135f660b8c947933c29eb8386ea8d8))
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20311/head:pull/20311` \
`$ git checkout pull/20311`

Update a local copy of the PR: \
`$ git checkout pull/20311` \
`$ git pull https://git.openjdk.org/jdk.git pull/20311/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20311`

View PR using the GUI difftool: \
`$ git pr show -t 20311`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20311.diff">https://git.openjdk.org/jdk/pull/20311.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20311#issuecomment-2247554554)

### CODE REVIEW

**Comment by TobiHartmann** [comment]
Github testing failed because [JDK-8336095](https://bugs.openjdk.org/browse/JDK-8336095) is not yet integrated. Testing with the fix all passed.

**Review by vnkozlov — APPROVED** [review]
Looks good to me.

**Comment by TobiHartmann** [comment]
Thanks for the review, Vladimir!

File: src/hotspot/share/opto/block.cpp [hunk 1]
```diff
@@ -39,6 +39,10 @@
 #include "utilities/powerOfTwo.hpp"
 
 void Block_Array::grow( uint i ) {
+  _nesting.check(_arena); // Check if a potential reallocation in the arena is safe
+  if (i < Max()) {
+    return; // No need to grow
+  }
   assert(i >= Max(), "must be an overflow");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/block.cpp:46]: The assert is now not necessary anymore. I guess you can remove it.

```suggestion
```

**Review by chhagedorn — APPROVED** [review]
Nice verification! Only one minor thing, otherwise, looks good to me, too.

**Review by chhagedorn — APPROVED** [review]
Still good!

**Pull request closed without merging**